File: android/app/src/main/java/com/rork/guidestreamtvandroid/data/remote/TrailerResolveService.kt
package com.rork.guidestreamtvandroid.data.remote

import com.rork.guidestreamtvandroid.SupabaseConfig
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.contentType
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

/**
 * Envelope for the `trailer_resolve` edge function response.
 */
@Serializable
data class TrailerResolveResponse(
    val ok: Boolean = false,
    val cached: Boolean = false,
    val keys: List<String> = emptyList(),
)

/**
 * Client for the `trailer_resolve` edge function (deployed with
 * verify_jwt=false, so the anon key is sufficient and no user session is
 * needed).
 *
 * TMDB reports which trailer keys exist for a title but never whether a key
 * will actually play (embedding may be owner-disabled, the video private, or
 * region-blocked). This server-side resolver verifies each candidate against
 * the YouTube Data API and returns only keys that are embeddable, public,
 * processed, and not US-blocked, in rank order. Uses the same raw Ktor
 * anon-key POST pattern as [WatchmodeResolveService].
 */
object TrailerResolveService {

    /**
     * Resolves verified playable YouTube trailer keys for a title.
     *
     * The nullable return is load-bearing and the two cases must never be
     * conflated:
     *  * Returns the decoded keys list on HTTP 200 — **including an empty
     *    list**, which means the title has no playable trailer at all (the
     *    caller drops it from the feed).
     *  * Returns null only when the call itself fails (a non-200 status or any
     *    caught exception), so the caller can degrade to the unverified TMDB
     *    key rather than emptying the feed.
     */
    suspend fun resolve(tmdbId: Int, isTV: Boolean): List<String>? {
        return try {
            val client = HttpClient {
                install(ContentNegotiation) { json(Json { ignoreUnknownKeys = true }) }
            }
            val url = "${SupabaseConfig.URL.trim()}/functions/v1/trailer_resolve"
            val body = buildJsonObject {
                put("tmdb_id", JsonPrimitive(tmdbId))
                put("media_type", JsonPrimitive(if (isTV) "tv" else "movie"))
            }
            val response: HttpResponse = client.post(url) {
                contentType(ContentType.Application.Json)
                header(HttpHeaders.ContentType, "application/json")
                header("apikey", SupabaseConfig.ANON_KEY)
                header(HttpHeaders.Authorization, "Bearer ${SupabaseConfig.ANON_KEY}")
                setBody(body.toString())
            }
            if (response.status.value == 200) {
                val resp: TrailerResolveResponse = response.body()
                resp.keys
            } else {
                null
            }
        } catch (_: Exception) {
            null
        }
    }
}

File: ios/GuideStreamTV/Services/TrailerResolveService.swift
//
//  TrailerResolveService.swift
//  GuideStreamTV
//
//  Client for the `trailer_resolve` Supabase edge function (deployed with
//  verify_jwt=false, so the anon key is sufficient and no user session is
//  needed).
//
//  TMDB reports which trailer keys exist for a title but never whether a key
//  will actually play: an owner may have disabled embedding, made the video
//  private, or region-blocked it. No client-side ranking can fix that because
//  the information isn't in the TMDB payload. The server-side resolver verifies
//  each candidate against the YouTube Data API and returns only keys that are
//  embeddable, public, processed, and not blocked in the US, in rank order — so
//  the Reels feed can trust the first key will play.
//
//  Uses the same anon-key POST header pattern as
//  `StreamAgentService.callAskStream`.
//

import Foundation

nonisolated enum TrailerResolveService {
    /// Shape returned by the `trailer_resolve` edge function.
    private struct Response: Decodable {
        let ok: Bool
        let cached: Bool
        let keys: [String]
    }

    /// Resolves the verified, playable YouTube trailer keys for a title in
    /// server rank order.
    ///
    /// The optional return is deliberately load-bearing and the two cases must
    /// never be conflated:
    ///  * Returns the decoded `keys` array on any HTTP 200 — **including an
    ///    empty array**, which is the server telling us this title has no
    ///    playable trailer at all (the caller drops it from the feed).
    ///  * Returns `nil` only when the call itself fails — a transport error, a
    ///    non-200 status, or a decode failure (the caller degrades to the
    ///    unverified TMDB key so a brief Supabase outage doesn't empty the feed).
    static func resolve(tmdbId: Int, isTV: Bool) async -> [String]? {
        let base = SupabaseConfig.url.trimmingCharacters(in: .whitespaces)
        guard let url = URL(string: "\(base)/functions/v1/trailer_resolve") else { return nil }

        let body: [String: Any] = [
            "tmdb_id": tmdbId,
            "media_type": isTV ? "tv" : "movie"
        ]
        guard let httpBody = try? JSONSerialization.data(withJSONObject: body) else { return nil }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(SupabaseConfig.anonKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(SupabaseConfig.anonKey)", forHTTPHeaderField: "Authorization")
        request.httpBody = httpBody

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            return nil
        }

        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else { return nil }
        guard let decoded = try? JSONDecoder().decode(Response.self, from: data) else { return nil }
        return decoded.keys
    }
}
